segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/BookKeeperStorageTest.java

```diff
@@ -0,0 +1,169 @@
+        ZK_CLIENT.set(CuratorFrameworkFactory
+                .builder()
+                .connectString("localhost:" + BK_PORT.get())
+                .namespace(namespace)
+                .retryPolicy(new ExponentialBackoffRetry(1000, 5))
+                .build());
+        ZK_CLIENT.get().start();
+
+        // Setup config to use the port and namespace.
+        CONFIG.set(BookKeeperStorageConfig
+                .builder()
+                .with(BookKeeperStorageConfig.ZK_ADDRESS, "localhost:" + BK_PORT.get())
+                .with(BookKeeperStorageConfig.ZK_METADATA_PATH, namespace)
+                .with(BookKeeperStorageConfig.BK_LEDGER_PATH, "/pravega/bookkeeper/ledgers")
+                .with(BookKeeperStorageConfig.BK_ENSEMBLE_SIZE, 1)
+                .with(BookKeeperStorageConfig.BK_WRITE_QUORUM_SIZE, 1)
+                .with(BookKeeperStorageConfig.BK_ACK_QUORUM_SIZE, 1)
+                .with(BookKeeperStorageConfig.BK_WRITE_TIMEOUT, 1000) // This is the minimum we can set anyway.
+                .build());
+    }
+
+    @AfterClass
+    public static void tearDownBookKeeper() throws Exception {
+        val process = BK_SERVICE.getAndSet(null);
+        if (process != null) {
+            process.close();
+        }
+    }
+
+    /**
+     * Tests fencing abilities. We create two different Storage objects with different owner ids.
+     * Part 1: Creation:
+     * * We create the Segment on Storage1:
+     * ** We verify that Storage1 can execute all operations.
+     * ** We verify that Storage2 can execute only read-only operations.
+     * * We open the Segment on Storage2:
+     * ** We verify that Storage1 can execute only read-only operations.
+     * ** We verify that Storage2 can execute all operations.
+     */
+    @Test
+    @Override
+    public void testFencing() throws Exception {
+        final long epoch1 = 1;
+        final long epoch2 = 2;
+        final String segmentName = "segment";
+        try (val storage1 = createStorage();
+             val storage2 = createStorage()) {
+            storage1.initialize(epoch1);
+            storage2.initialize(epoch2);
+
+            // Create segment in Storage1 (thus Storage1 owns it for now).
+            storage1.create(segmentName, TIMEOUT).join();
+
+            // Storage1 should be able to execute all operations.
+            SegmentHandle handle1 = storage1.openWrite(segmentName).join();
+            verifyWriteOperationsSucceed(handle1, storage1);
+            verifyReadOnlyOperationsSucceed(handle1, storage1);
+
+            // Open the segment in Storage2 (thus Storage2 owns it for now).
+            SegmentHandle handle2 = storage2.openWrite(segmentName).join();
+
+            // Storage1 should be able to execute only read-only operations.
+            verifyWriteOperationsFail(handle1, storage1);
+            //verifyReadOnlyOperationsSucceed(handle1, storage1);
```

Comment:
This shouldn't be commented out, yes?

Reply:
This behavior of `BookKeeperStorage` is different than other storage. It depends on the fact that read and write operation always end up on the same segmentstore which owns the container that the segment belongs to.
This helps to optimize the ZK usage. The alternative is that all the storages that are interested in this segment keep the ZK data in sync. This will be costly and unnecessary.
The modified behavior is that the segmentstores which do not own the segment can read optimistically from the segment but may or may not not get the latest data.